File: csharp/BronKerboschUnitTest/UtilUnitTest.cs
using NUnit.Framework;
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using Vertex = System.UInt32;

namespace BronKerboschUnitTest
{
    public class UtilTest
    {
        [Test]
        public void Util_Append()
        {
            var empty = ImmutableArray.Create<Vertex>();
            var one = CollectionsUtil.Append(empty, 11);
            var two = CollectionsUtil.Append(one, 22);
            Assert.That(empty.Length.Equals(0));
            Assert.That(one.Length.Equals(1));
            Assert.That(two.Length.Equals(2));
            Assert.That(one[0].Equals(11u));
            Assert.That(two[0].Equals(11u));
            Assert.That(two[1].Equals(22u));
        }

        [Test]
        public void Util_PopArbitrary()
        {
            var one = new HashSet<Vertex> { 1 };
            var two = new HashSet<Vertex> { 1, 2 };
            Assert.That(CollectionsUtil.PopArbitrary(one).Equals(1u));
            Assert.Zero(one.Count());
            var x = CollectionsUtil.PopArbitrary(two);
            var y = CollectionsUtil.PopArbitrary(two);
            Assert.That(Math.Min(x, y).Equals(1u));
            Assert.That(Math.Max(x, y).Equals(2u));
        }

        [Test]
        public void Util_AreDisjoint()
        {
            var empty = new HashSet<Vertex> { };
            var one = new HashSet<Vertex> { 1 };
            var two = new HashSet<Vertex> { 1, 2 };
            var six = new HashSet<Vertex> { 0, 1, 2, 3, 4, 5 };
            Assert.That(CollectionsUtil.AreDisjoint(empty, one));
            Assert.That(CollectionsUtil.AreDisjoint(one, empty));
            Assert.That(CollectionsUtil.AreDisjoint(empty, two));
            Assert.That(CollectionsUtil.AreDisjoint(two, empty));
            Assert.That(CollectionsUtil.AreDisjoint(empty, six));
            Assert.That(CollectionsUtil.AreDisjoint(six, empty));
            Assert.That(!CollectionsUtil.AreDisjoint(one, two));
            Assert.That(!CollectionsUtil.AreDisjoint(two, one));
            Assert.That(!CollectionsUtil.AreDisjoint(one, six));
            Assert.That(!CollectionsUtil.AreDisjoint(six, one));
            Assert.That(!CollectionsUtil.AreDisjoint(two, six));
            Assert.That(!CollectionsUtil.AreDisjoint(six, two));
            Assert.That(!CollectionsUtil.AreDisjoint(one, one));
            Assert.That(!CollectionsUtil.AreDisjoint(two, two));
            Assert.That(!CollectionsUtil.AreDisjoint(six, six));
        }

        [Test]
        public void Util_Intersection()
        {
            var empty = new HashSet<Vertex> { };
            var one = new HashSet<Vertex> { 1 };
            var two = new HashSet<Vertex> { 1, 2 };
            var six = new HashSet<Vertex> { 0, 1, 2, 3, 4, 5 };
            Assert.That(CollectionsUtil.Intersection(empty, one).SetEquals(empty));
            Assert.That(CollectionsUtil.Intersection(one, empty).SetEquals(empty));
            Assert.That(CollectionsUtil.Intersection(empty, two).SetEquals(empty));
            Assert.That(CollectionsUtil.Intersection(two, empty).SetEquals(empty));
            Assert.That(CollectionsUtil.Intersection(empty, six).SetEquals(empty));
            Assert.That(CollectionsUtil.Intersection(six, empty).SetEquals(empty));
            Assert.That(CollectionsUtil.Intersection(one, two).SetEquals(one));
            Assert.That(CollectionsUtil.Intersection(two, one).SetEquals(one));
            Assert.That(CollectionsUtil.Intersection(one, six).SetEquals(one));
            Assert.That(CollectionsUtil.Intersection(six, one).SetEquals(one));
            Assert.That(CollectionsUtil.Intersection(two, six).SetEquals(two));
            Assert.That(CollectionsUtil.Intersection(six, two).SetEquals(two));
            Assert.That(CollectionsUtil.Intersection(one, one).SetEquals(one));
            Assert.That(CollectionsUtil.Intersection(two, two).SetEquals(two));
            Assert.That(CollectionsUtil.Intersection(six, six).SetEquals(six));
        }

        [Test]
        public void Util_IntersectCount()
        {
            var empty = new HashSet<Vertex> { };
            var one = new HashSet<Vertex> { 1 };
            var two = new HashSet<Vertex> { 1, 2 };
            var six = new HashSet<Vertex> { 0, 1, 2, 3, 4, 5 };
            Assert.That(CollectionsUtil.IntersectionSize(empty, one).Equals(0));
            Assert.That(CollectionsUtil.IntersectionSize(one, empty).Equals(0));
            Assert.That(CollectionsUtil.IntersectionSize(empty, two).Equals(0));
            Assert.That(CollectionsUtil.IntersectionSize(two, empty).Equals(0));
            Assert.That(CollectionsUtil.IntersectionSize(empty, six).Equals(0));
            Assert.That(CollectionsUtil.IntersectionSize(six, empty).Equals(0));
            Assert.That(CollectionsUtil.IntersectionSize(one, two).Equals(1));
            Assert.That(CollectionsUtil.IntersectionSize(two, one).Equals(1));
            Assert.That(CollectionsUtil.IntersectionSize(one, six).Equals(1));
            Assert.That(CollectionsUtil.IntersectionSize(six, one).Equals(1));
            Assert.That(CollectionsUtil.IntersectionSize(two, six).Equals(2));
            Assert.That(CollectionsUtil.IntersectionSize(six, two).Equals(2));
            Assert.That(CollectionsUtil.IntersectionSize(one, one).Equals(1));
            Assert.That(CollectionsUtil.IntersectionSize(two, two).Equals(2));
            Assert.That(CollectionsUtil.IntersectionSize(six, six).Equals(6));
        }

        [Test]
        public void Util_Difference()
        {
            var empty = new HashSet<Vertex> { };
            var one = new HashSet<Vertex> { 1 };
            var two = new HashSet<Vertex> { 1, 2 };
            var six = new HashSet<Vertex> { 0, 1, 2, 3, 4, 5 };
            Assert.That(CollectionsUtil.Difference(empty, one).SetEquals(empty));
            Assert.That(CollectionsUtil.Difference(empty, two).SetEquals(empty));
            Assert.That(CollectionsUtil.Difference(empty, six).SetEquals(empty));
            Assert.That(CollectionsUtil.Difference(one, one).SetEquals(empty));
            Assert.That(CollectionsUtil.Difference(one, two).SetEquals(empty));
            Assert.That(CollectionsUtil.Difference(one, six).SetEquals(empty));
            Assert.That(CollectionsUtil.Difference(two, two).SetEquals(empty));
            Assert.That(CollectionsUtil.Difference(two, six).SetEquals(empty));
            Assert.That(CollectionsUtil.Difference(six, six).SetEquals(empty));
            Assert.That(CollectionsUtil.Difference(one, empty).SetEquals(one));
            Assert.That(CollectionsUtil.Difference(two, empty).SetEquals(two));
            Assert.That(CollectionsUtil.Difference(six, empty).SetEquals(six));
            Assert.That(CollectionsUtil.Difference(two, one).SetEquals(new HashSet<Vertex> { 2 }));
            Assert.That(CollectionsUtil.Difference(six, one).SetEquals(new HashSet<Vertex> { 0, 2, 3, 4, 5 }));
            Assert.That(CollectionsUtil.Difference(six, two).SetEquals(new HashSet<Vertex> { 0, 3, 4, 5 }));
        }
    }
}

File: cpp/BronKerboschUnitTest/UtilUnitTest.cpp
#include "pch.h"

#include "BronKerbosch/Util.h"

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace BronKerbosch {
    TEST_CLASS(UtilUnitTest) {
       public:
        template <typename VertexSet>
        void Util_pop_arbitray() {
            auto one = VertexSet{1};
            auto two = VertexSet{1, 2};
            Assert::AreEqual(1, Util::pop_arbitrary(one));
            Assert::IsTrue(one.empty());
            auto x = Util::pop_arbitrary(two);
            auto y = Util::pop_arbitrary(two);
            Assert::AreEqual(1, std::min(x, y));
            Assert::AreEqual(2, std::max(x, y));
        }

        template <typename VertexSet>
        void Util_are_disjoint() {
            auto const empty = VertexSet{};
            auto const one = VertexSet{1};
            auto const two = VertexSet{1, 2};
            auto const six = VertexSet{0, 1, 2, 3, 4, 5};
            Assert::IsTrue(Util::are_disjoint(empty, one));
            Assert::IsTrue(Util::are_disjoint(one, empty));
            Assert::IsTrue(Util::are_disjoint(empty, two));
            Assert::IsTrue(Util::are_disjoint(two, empty));
            Assert::IsTrue(Util::are_disjoint(empty, six));
            Assert::IsTrue(Util::are_disjoint(six, empty));
            Assert::IsFalse(Util::are_disjoint(one, two));
            Assert::IsFalse(Util::are_disjoint(two, one));
            Assert::IsFalse(Util::are_disjoint(one, six));
            Assert::IsFalse(Util::are_disjoint(six, one));
            Assert::IsFalse(Util::are_disjoint(two, six));
            Assert::IsFalse(Util::are_disjoint(six, two));
            Assert::IsFalse(Util::are_disjoint(one, one));
            Assert::IsFalse(Util::are_disjoint(two, two));
            Assert::IsFalse(Util::are_disjoint(six, six));
        }

        template <typename VertexSet>
        void Util_intersection() {
            auto const empty = VertexSet{};
            auto const one = VertexSet{1};
            auto const two = VertexSet{1, 2};
            auto const six = VertexSet{0, 1, 2, 3, 4, 5};
            Assert::IsTrue(Util::intersection(empty, one) == empty);
            Assert::IsTrue(Util::intersection(one, empty) == empty);
            Assert::IsTrue(Util::intersection(empty, two) == empty);
            Assert::IsTrue(Util::intersection(two, empty) == empty);
            Assert::IsTrue(Util::intersection(empty, six) == empty);
            Assert::IsTrue(Util::intersection(six, empty) == empty);
            Assert::IsTrue(Util::intersection(one, two) == one);
            Assert::IsTrue(Util::intersection(two, one) == one);
            Assert::IsTrue(Util::intersection(one, six) == one);
            Assert::IsTrue(Util::intersection(six, one) == one);
            Assert::IsTrue(Util::intersection(two, six) == two);
            Assert::IsTrue(Util::intersection(six, two) == two);
            Assert::IsTrue(Util::intersection(one, one) == one);
            Assert::IsTrue(Util::intersection(two, two) == two);
            Assert::IsTrue(Util::intersection(six, six) == six);
        }

        template <typename VertexSet>
        void Util_intersection_size() {
            auto const empty = VertexSet{};
            auto const one = VertexSet{1};
            auto const two = VertexSet{1, 2};
            auto const six = VertexSet{0, 1, 2, 3, 4, 5};
            Assert::IsTrue(Util::intersection_size(empty, one) == 0);
            Assert::IsTrue(Util::intersection_size(one, empty) == 0);
            Assert::IsTrue(Util::intersection_size(empty, two) == 0);
            Assert::IsTrue(Util::intersection_size(two, empty) == 0);
            Assert::IsTrue(Util::intersection_size(empty, six) == 0);
            Assert::IsTrue(Util::intersection_size(six, empty) == 0);
            Assert::IsTrue(Util::intersection_size(one, two) == 1);
            Assert::IsTrue(Util::intersection_size(two, one) == 1);
            Assert::IsTrue(Util::intersection_size(one, six) == 1);
            Assert::IsTrue(Util::intersection_size(six, one) == 1);
            Assert::IsTrue(Util::intersection_size(two, six) == 2);
            Assert::IsTrue(Util::intersection_size(six, two) == 2);
            Assert::IsTrue(Util::intersection_size(one, one) == 1);
            Assert::IsTrue(Util::intersection_size(two, two) == 2);
            Assert::IsTrue(Util::intersection_size(six, six) == 6);
        }

        template <typename VertexSet>
        void Util_difference() {
            auto const empty = VertexSet{};
            auto const one = VertexSet{1};
            auto const two = VertexSet{1, 2};
            auto const six = VertexSet{0, 1, 2, 3, 4, 5};
            Assert::IsTrue(Util::difference(empty, one) == empty);
            Assert::IsTrue(Util::difference(empty, two) == empty);
            Assert::IsTrue(Util::difference(empty, six) == empty);
            Assert::IsTrue(Util::difference(one, one) == empty);
            Assert::IsTrue(Util::difference(one, two) == empty);
            Assert::IsTrue(Util::difference(one, six) == empty);
            Assert::IsTrue(Util::difference(two, two) == empty);
            Assert::IsTrue(Util::difference(two, six) == empty);
            Assert::IsTrue(Util::difference(six, six) == empty);
            Assert::IsTrue(Util::difference(one, empty) == one);
            Assert::IsTrue(Util::difference(two, empty) == two);
            Assert::IsTrue(Util::difference(six, empty) == six);
            Assert::IsTrue(Util::difference(two, one) == VertexSet{2});
            Assert::IsTrue(Util::difference(six, one) == VertexSet{0, 2, 3, 4, 5});
            Assert::IsTrue(Util::difference(six, two) == VertexSet{0, 3, 4, 5});
        }

        TEST_METHOD(Util_pop_arbitray_set) {
            Util_pop_arbitray<std::set<int>>();
        }
        TEST_METHOD(Util_pop_arbitray_ordered_vector) {
            Util_pop_arbitray<ordered_vector<int>>();
        }
        TEST_METHOD(Util_pop_arbitray_unordered_set) {
            Util_pop_arbitray<std::unordered_set<int>>();
        }

        TEST_METHOD(Util_are_disjoint_set) {
            Util_are_disjoint<std::set<int>>();
        }
        TEST_METHOD(Util_are_disjoint_ordered_vector) {
            Util_are_disjoint<ordered_vector<int>>();
        }
        TEST_METHOD(Util_are_disjoint_unordered_set) {
            Util_are_disjoint<std::unordered_set<int>>();
        }

        TEST_METHOD(Util_intersection_size_set) {
            Util_intersection_size<std::set<int>>();
        }
        TEST_METHOD(Util_intersection_size_ordered_vector) {
            Util_intersection_size<ordered_vector<int>>();
        }
        TEST_METHOD(Util_intersection_size_unordered_set) {
            Util_intersection_size<std::unordered_set<int>>();
        }

        TEST_METHOD(Util_intersection_set) {
            Util_intersection<std::set<int>>();
        }
        TEST_METHOD(Util_intersection_ordered_vector) {
            Util_intersection<ordered_vector<int>>();
        }
        TEST_METHOD(Util_intersection_unordered_set) {
            Util_intersection<std::unordered_set<int>>();
        }

        TEST_METHOD(Util_difference_set) {
            Util_difference<std::set<int>>();
        }
        TEST_METHOD(Util_difference_ordered_vector) {
            Util_difference<ordered_vector<int>>();
        }
        TEST_METHOD(Util_difference_unordered_set) {
            Util_difference<std::unordered_set<int>>();
        }
    };
}
